test/spec/tokenManager.js

```diff
@@ -1513,38 +1514,136 @@ describe('TokenManager', function() {
-      setupSync();
-      client.tokenManager.add = jest.fn();
-      client.tokenManager.setTokens({ 
-        accessToken: tokens.standardAccessTokenParsed
-      });
-      oauthUtil.expectTokenStorageToEqual(localStorage, {
-        'accessToken': tokens.standardAccessTokenParsed
+      setupSync({
+        tokenManager: {
+          storage: storageProvider
+        }
       });
+      const tokensObj = { 
+        accessToken: tokens.standardAccessTokenParsed, 
+      };
+      client.tokenManager.setTokens(tokensObj);
+      expect(setItemMock).toHaveBeenCalledTimes(1);
+      expect(setItemMock).toHaveBeenCalledWith('okta-token-storage', JSON.stringify(tokensObj));
     });
 
     it('should add idToken to storage', () => {
-      setupSync();
-      client.tokenManager.add = jest.fn();
-      client.tokenManager.setTokens({ 
-        idToken: tokens.standardIdTokenParsed
+      setupSync({
+        tokenManager: {
+          storage: storageProvider
+        }
       });
-      oauthUtil.expectTokenStorageToEqual(localStorage, {
-        'idToken': tokens.standardIdTokenParsed
+      const tokensObj = { 
+        idToken: tokens.standardIdTokenParsed,
+      };
+      client.tokenManager.setTokens(tokensObj);
+      expect(setItemMock).toHaveBeenCalledTimes(1);
+      expect(setItemMock).toHaveBeenCalledWith('okta-token-storage', JSON.stringify(tokensObj));
+    });
+  });
+
+  describe('cross tabs communication', () => {
+    let sdkMock;
+    beforeEach(function() {
+      const emitter = new Emitter();
+      sdkMock = {
+        options: {},
+        emitter
+      };
+    });
+    it('should emit events and reset timeouts when storage event happen with token storage key', () => {
+      const instance = new TokenManager(sdkMock);
+      instance._resetExpireEventTimeoutAll = jest.fn();
+      instance._emitEventsForCrossTabsStorageUpdate = jest.fn();
+      window.dispatchEvent(new StorageEvent('storage', {
+        key: 'okta-token-storage', 
+        newValue: 'fake_new_value',
+        oldValue: 'fake_old_value'
+      }));
+      expect(instance._resetExpireEventTimeoutAll).toHaveBeenCalled();
+      expect(instance._emitEventsForCrossTabsStorageUpdate).toHaveBeenCalledWith('fake_new_value', 'fake_old_value');
+    });
+    it('should not emit events or reset timeouts if the key is not token storage key', () => {
+      const instance = new TokenManager(sdkMock);
+      instance._resetExpireEventTimeoutAll = jest.fn();
+      instance._emitEventsForCrossTabsStorageUpdate = jest.fn();
+      window.dispatchEvent(new StorageEvent('storage', {
+        key: 'fake-key', 
+        newValue: 'fake_new_value',
+        oldValue: 'fake_old_value'
+      }));
+      expect(instance._resetExpireEventTimeoutAll).not.toHaveBeenCalled();
+      expect(instance._emitEventsForCrossTabsStorageUpdate).not.toHaveBeenCalled();
+    });
+
+    describe('_emitEventsForCrossTabsStorageUpdate', () => {
```

Comment:
Is there a way to test to make sure that we don't cause extra calls?  I don't see anything in the code that's wrong, but a test making sure that an emitted event doesn't cause a chain of events continuously bouncing between tabs would help future-proof the code.

Reply:
Extra calls may happen, as we still want to keep autoRenew running in each single tab. Then we handle it in AuthStateManager to throttle the emitted events by comparing the state and cancelablePromise. https://github.com/okta/okta-auth-js/pull/484/files#diff-c87977be240b146c8cc1ab3c11a6cc66R71
Also, the events just triggered by storage change, but it will not trigger another storage change, then we should be save from the infinite loop.

Reply:
Overall, yes, extra calls may happen.  I'm just saying we should test to make sure that if only one call is expected (because artificial test situation) only one call happens.

That helps make sure we don't shoot ourselves in the foot later (as you say, currently it won't happen.  I want to make sure later changes don't break that assumption).

Reply:
I agree with the `future-proof` point. I just don't feel it's easy to simulate the multiple tabs env in unit test, if we just trigger the function by emitting a StorageEvent, I don't think it can help too much.

Reply:
Hmm, that's more involved than I was thinking.  More of:
- test that when we make a change, we only emit exactly one event, not more than one
- test that when we get an event that doesn't require a change, we don't emit an event

Reply:
StorageEvent can only be triggered outside the current tab, maybe I can try `wdio ` as Aaron suggested.

> test that when we get an event that doesn't require a change, we don't emit an event

yep 👍 , looks like more tests are needed.